HTML: Handle shadow host with delegatesFocus=true in Element.focus(), click and sequential focus navigation

rakina opened the pull request:
WPT for https://github.com/whatwg/html/pull/4796.
As noted there, the behavior in Chrome for focus() when the flattened tabindex-ordered focus navigation scope is empty (there's nothing to focus on) so it fails in some of the tests.

rakina committed:
HTML: Handle shadow host with delegatesFocus=true in Element.focus()/focusing steps

rniwa commented on shadow-dom/focus/focus-method-delegatesFocus.html:
```diff
@@ -0,0 +1,289 @@
+  // flattened tabindex-ordered focus navigation scope of #host: []
+  host.focus();
+  assert_equals(shadowRoot.activeElement, null);
+  assert_equals(document.activeElement, null);
```
How does this expectation make sense? `resetFocus()` call above just moved the focus to `document.body`. There is no focusable area within this shadow tree so we should be bailing out before moving any focus so `host.focus()` should really be a no-op.

Also, the current definition makes it so that `document.activeElement` almost never returns `null` so long as the document has body or a document element:
https://html.spec.whatwg.org/multipage/interaction.html#dom-document-activeelement

rniwa replied:
So I think you'd have to replace everywhere you assert `document.activeElement` to be `null` to assert it's `document.body`. Then all test cases would pass in WebKit once I've implemented all the relevant bits and pieces.

rakina replied:
`resetFocus()` calls `document.activeElement.blur()` which I assume would remove any existing focus in the document? At least in Blink that means `document.activeElement` would be null after. Is it not supposed to be like that?

rniwa replied:
That's odd. The version of `shadow-utils.js` I have is as follows:
```js
function resetFocus() {
  document.body.focus();
}
```

rakina replied:
Aah I see, yeah this test file doesn't use `shadow-utils.js` at all. It uses the resetFocus() defined in line 67. Since the PR for the sequential focus WPT is merged already https://github.com/web-platform-tests/wpt/pull/17523, should I make these tests use shadow-utils as well? (or maybe I should just rename shadow-utils to sequential-focus-utils..)

rakina replied:
Updated the tests, PTAL - also added sequential focus navigation tests.

rakina committed:
Update focus() test, add sequential focus navigation tests

rakina commented:
Updated the tests after the change in https://github.com/whatwg/html/pull/4796. We have tests for focus() and sequential focus navigation here, since click focus delegation can't really be tested? (there is no difference between focusable and click focusable in Blink)

domenic commented on shadow-dom/focus/focus-tabindex-order-shadow-varying-delegatesFocus.html:
```diff
@@ -0,0 +1,36 @@
+  setTabIndex([slotBelow], 2);
+  setTabIndex([slottedAbove], 3);
+  resetFocus();
+  return assertFocusOrder([aboveHost, slottedAbove, slottedBelow, aboveSlot, belowHost]);
```
Why is this slottedAbove, slottedBelow, aboveSlot, instead of the reverse order? The tabindex goes aboveSlot = 0, slottedBelow = 1, slottedAbove = 3, right?

rakina replied:
tabindex order is 1,2,3,...,0 so #aboveSlot will be last.
construction of flattened tabindex-ordered focus navigation scope of #host: 
[#slotAbove (tabindex=1), #slotBelow (tabindex=2), #aboveSlot (tabindex=0)]
replace #slotAbove and #slotBelow with its contents:
[#slottedAbove (tabindex=3), #slottedBelow (tabindex=1), #aboveSlot (tabindex=0)]

domenic replied:
Ah right, OK, makes sense!

domenic commented on shadow-dom/focus/focus-tabindex-order-shadow-zero-delegatesFocus.html:
```diff
@@ -0,0 +1,34 @@
+  resetFocus();
+  // Focus should move in flat tree order since every one of them has tabindex ==0,
+  // but doesn't include slots and #host.
+  return assertFocusOrder(elementsInFlatTreeOrder.filter(el => (el !== slotAbove && el !== slotBelow && el !== host)));
```
I think listing them explicitly would make the test easier to read...

domenic reviewed (commented):
Very nice exhaustive tests :D.

We should probably also add tests for .click(), I think... in particular checking that it doesn't use sequential focus navigation order, in the way Chrome does currently.

domenic reviewed (approved):
LGTM with nit about explicitly listing them

rakina committed:
Explicitly list order

rakina commented:
> We should probably also add tests for .click(), I think... in particular checking that it doesn't use sequential focus navigation order, in the way Chrome does currently.

`click()` fires a synthetic click event, which triggers activation behavior but not focus - discovered this with @tkent-google earlier today, gotta update that part of the spec D: - so we can't test click focus with that. I'm not sure if there's another way to trigger clicks in WPTs...

tkent-google commented:
>  I'm not sure if there's another way to trigger clicks in WPTs...

You can do it by `test_driver.click(element)`.

rakina commented:
> > I'm not sure if there's another way to trigger clicks in WPTs...
> 
> You can do it by `test_driver.click(element)`.

Tried this just now, looks like this also doesn't trigger the focus on any element I tried it on :(

domenic commented:
Let's at least add a basic test or two that `.click()` does not focus things, since we also need to update that in the PR.

rakina committed:
Add test for click focus

rakina commented:
> Let's at least add a basic test or two that `.click()` does not focus things, since we also need to update that in the PR.

Makes sense, added. Thanks!

domenic commented on shadow-dom/focus/click-focus-delegatesFocus.html:
```diff
@@ -0,0 +1,83 @@
+test(() => {
+  resetTabIndexAndFocus();
+  setAllTabIndex(0);
+  test_driver.click(host);
```
I don't think we should have these test_driver tests. It feels like a bug in test_driver that it doesn't properly emulate real clicks. So, this test is really a test of test_driver, not a test of the spec, and I think it's testing that test_driver has a bug.

rakina replied:
Just found out that test_driver.click actually works (click() still won't focus though), so I separated the test driver tests now. PTAL!

tkent-google commented on shadow-dom/focus/click-focus-delegatesFocus.html:
```diff
@@ -0,0 +1,83 @@
+test(() => {
+  resetTabIndexAndFocus();
+  setAllTabIndex(0);
+  test_driver.click(host);
```
test_driver.click() is asynchronous.  We need to wait until the retuned promise is resolved.

```
promise_test(async () => {
  reset...
  setAll...
  await test_driver.click(host);
  assert_equals(...
  assert_equals(...
}, ...

```

rakina replied:
OMG, you're right. It actually focuses now!! Thank you :D

tkent-google commented on shadow-dom/focus/click-focus-delegatesFocus.html:
```diff
@@ -0,0 +1,83 @@
+test(() => {
+  resetTabIndexAndFocus();
+  setAllTabIndex(0);
+  test_driver.click(slotted);
```
Ditto.

rakina committed:
Add actual testdriver tests

domenic commented on shadow-dom/focus/click-focus-delegatesFocus-tabindex-varies.html:
```diff
@@ -0,0 +1,69 @@
+  resetFocus(shadowRoot);
+}
+
+async_test(async (t) => {
```
Best to use promise_test, then you do not need t.step or t.done

rakina replied:
Done!

rakina committed:
use promise_test

rakina committed:
Add tabindex=1 test to focus()

We should use flat-tree order instead of tabindex order, add test to ensure that.

rniwa commented on shadow-dom/focus/click-focus-delegatesFocus-tabindex-varies.html:
```diff
@@ -0,0 +1,68 @@
+  setTabIndex([aboveSlot], 2);
+  setTabIndex([slot, slotted], 1);
+  await test_driver.click(host);
+  assert_equals(shadowRoot.activeElement, null);
```
Clearly you meant aboveSlot here.

rakina replied:
Oops, thanks!

rniwa commented on shadow-dom/focus/click-focus-delegatesFocus-tabindex-varies.html:
```diff
@@ -0,0 +1,68 @@
+  setTabIndex([slot, slotted], 1);
+  await test_driver.click(host);
+  assert_equals(shadowRoot.activeElement, null);
+  assert_equals(document.activeElement, aboveSlot);
```
And host here.

rakina replied:
Done.

rakina committed:
Update focus-method-delegatesFocus.html

rakina committed:
Update click-focus-delegatesFocus-tabindex-varies.html

rakina commented on shadow-dom/focus/focus-method-delegatesFocus.html:
```diff
@@ -0,0 +1,232 @@
+  // No focusable element under #host in the flat tree.
+  host.focus();
+  assert_equals(shadowRoot.activeElement, null);
+  assert_equals(document.activeElement, null);
```
Done.

rakina commented on shadow-dom/focus/focus-method-delegatesFocus.html:
```diff
@@ -0,0 +1,232 @@
+  // No focusable element under #host in the flat tree.
+  host.focus();
+  assert_equals(shadowRoot.activeElement, null);
+  assert_equals(document.activeElement, null);
```
Done.

rakina reviewed (commented):
Thanks for the review @rniwa!

rniwa reviewed (approved):
r=me.
